Gap Shuffler

Chris3606 opened the pull request:
This PR implements a "gap-shuffle" algorithm heavily based on SquidSquad's GapShuffler.  It functions on IEnumerable, IList, Spans, and Memory representations of a list, and offers methods to shuffle an existing list in-place rather than copying its elements.

It uses a custom enumerator struct to ensure the implementation is as fast as reasonably possible when used in a foreach loop directly, and should still offer as reasonable as possible performance when used as an IEnumerable/with LINQ.

There are performance tests written, here are the most current results:

|                        Method | Size |        Mean |     Error |    StdDev |
|------------------------------ |----- |------------:|----------:|----------:|
|             ShaiRandomForeach |   10 |    181.4 ns |   0.64 ns |   0.50 ns |
|      ShaiRandomForeachInPlace |   10 |    153.1 ns |   0.64 ns |   0.57 ns |
|         ShaiRandomForeachTake |   10 |    329.7 ns |   2.50 ns |   2.21 ns |
|  ShaiRandomForeachInPlaceTake |   10 |    299.5 ns |   2.60 ns |   2.30 ns |
|            YieldReturnForeach |   10 |    253.7 ns |   2.61 ns |   2.44 ns |
|     YieldReturnForeachInPlace |   10 |    227.1 ns |   1.49 ns |   1.25 ns |
|        YieldReturnForeachTake |   10 |    339.6 ns |   1.76 ns |   1.56 ns |
| YieldReturnForeachInPlaceTake |   10 |    299.4 ns |   3.13 ns |   2.78 ns |
|                ShaiRandomLinq |   10 |    339.4 ns |   1.89 ns |   1.77 ns |
|         ShaiRandomLinqInPlace |   10 |    301.8 ns |   2.78 ns |   2.60 ns |
|               YieldReturnLinq |   10 |    325.4 ns |   1.05 ns |   0.82 ns |
|        YieldReturnLinqInPlace |   10 |    299.2 ns |   1.38 ns |   1.22 ns |
|         ShaiRandomIEnumerable |   10 |    227.8 ns |   1.15 ns |   0.96 ns |
|  ShaiRandomIEnumerableInPlace |   10 |    194.6 ns |   0.79 ns |   0.66 ns |
|             ShaiRandomForeach |   50 |    896.9 ns |   9.86 ns |   9.22 ns |
|      ShaiRandomForeachInPlace |   50 |    790.5 ns |   8.62 ns |   8.06 ns |
|         ShaiRandomForeachTake |   50 |  1,278.4 ns |   6.72 ns |   5.61 ns |
|  ShaiRandomForeachInPlaceTake |   50 |  1,238.2 ns |   5.49 ns |   4.87 ns |
|            YieldReturnForeach |   50 |  1,062.6 ns |   8.04 ns |   7.53 ns |
|     YieldReturnForeachInPlace |   50 |  1,067.0 ns |  11.19 ns |   9.92 ns |
|        YieldReturnForeachTake |   50 |  1,381.7 ns |   6.87 ns |   5.74 ns |
| YieldReturnForeachInPlaceTake |   50 |  1,284.3 ns |  10.25 ns |   9.09 ns |
|                ShaiRandomLinq |   50 |  1,312.6 ns |   4.44 ns |   4.15 ns |
|         ShaiRandomLinqInPlace |   50 |  1,234.7 ns |   3.99 ns |   3.54 ns |
|               YieldReturnLinq |   50 |  1,307.8 ns |   7.15 ns |   6.34 ns |
|        YieldReturnLinqInPlace |   50 |  1,258.1 ns |   7.36 ns |   6.88 ns |
|         ShaiRandomIEnumerable |   50 |    954.5 ns |   5.72 ns |   5.07 ns |
|  ShaiRandomIEnumerableInPlace |   50 |    901.9 ns |   6.17 ns |   5.47 ns |
|             ShaiRandomForeach |  100 |  1,771.8 ns |   7.08 ns |   6.62 ns |
|      ShaiRandomForeachInPlace |  100 |  1,580.3 ns |  12.28 ns |  10.25 ns |
|         ShaiRandomForeachTake |  100 |  2,449.0 ns |  26.14 ns |  24.45 ns |
|  ShaiRandomForeachInPlaceTake |  100 |  2,420.2 ns |  25.36 ns |  23.72 ns |
|            YieldReturnForeach |  100 |  2,100.5 ns |  24.74 ns |  23.15 ns |
|     YieldReturnForeachInPlace |  100 |  2,119.5 ns |  31.29 ns |  26.13 ns |
|        YieldReturnForeachTake |  100 |  2,568.7 ns |  18.23 ns |  16.16 ns |
| YieldReturnForeachInPlaceTake |  100 |  2,533.0 ns |  15.50 ns |  12.94 ns |
|                ShaiRandomLinq |  100 |  2,530.9 ns |  13.15 ns |  10.98 ns |
|         ShaiRandomLinqInPlace |  100 |  2,477.2 ns |  33.98 ns |  31.78 ns |
|               YieldReturnLinq |  100 |  2,593.8 ns |   8.45 ns |   6.60 ns |
|        YieldReturnLinqInPlace |  100 |  2,469.6 ns |   9.95 ns |   8.82 ns |
|         ShaiRandomIEnumerable |  100 |  1,938.1 ns |  33.57 ns |  28.03 ns |
|  ShaiRandomIEnumerableInPlace |  100 |  1,796.9 ns |  13.76 ns |  12.20 ns |
|             ShaiRandomForeach |  500 |  8,069.3 ns |  71.43 ns |  66.82 ns |
|      ShaiRandomForeachInPlace |  500 |  7,868.1 ns |  49.66 ns |  46.46 ns |
|         ShaiRandomForeachTake |  500 | 11,744.7 ns |  56.08 ns |  52.46 ns |
|  ShaiRandomForeachInPlaceTake |  500 | 12,007.3 ns |  34.62 ns |  30.69 ns |
|            YieldReturnForeach |  500 | 10,235.0 ns |  46.98 ns |  41.65 ns |
|     YieldReturnForeachInPlace |  500 | 10,582.6 ns |  66.67 ns |  55.67 ns |
|        YieldReturnForeachTake |  500 | 12,480.8 ns |  93.55 ns |  82.93 ns |
| YieldReturnForeachInPlaceTake |  500 | 12,606.0 ns |  62.80 ns |  58.74 ns |
|                ShaiRandomLinq |  500 | 12,512.6 ns | 106.35 ns |  94.28 ns |
|         ShaiRandomLinqInPlace |  500 | 12,044.0 ns | 107.10 ns |  94.94 ns |
|               YieldReturnLinq |  500 | 12,634.3 ns | 118.48 ns | 110.83 ns |
|        YieldReturnLinqInPlace |  500 | 12,170.7 ns |  74.08 ns |  69.29 ns |
|         ShaiRandomIEnumerable |  500 |  9,352.5 ns |  85.16 ns |  79.66 ns |
|  ShaiRandomIEnumerableInPlace |  500 |  8,787.7 ns |  48.35 ns |  45.22 ns |

Note that the custom enumerator seems to out-perform a more traditional "yield return" implementation considerably.  The use of System.LINQ does slow it down notably, but not much compared to the equivalent operations with `yield return`.

Chris3606 committed:
Added basic gap shuffling implementation.

Chris3606 committed:
Added performance tests to verify the GapShuffle custom enumerator is the way to go.

Chris3606 committed:
Documented methods and optimized enumerator by not using .Count as frequently.

Chris3606 committed:
Added versions of GapShuffleInPlace for Memory<T>.

tommyettinger commented:
I'll take a look! I just want to summarize what GapShuffler is in case anyone reading this later isn't familiar. This also might be handy for documentation, if anything needs more.

GapShuffler (in SquidLib or SquidSquad) is akin to an infinite-length IEnumerable that shuffles a sequence, iterates over the shuffled elements as they are requested, and when it runs out of elements it shuffles the sequence again. The Gap in the name refers to how it prevents the most-recently returned item in the sequence from being returned again immediately after the items are shuffled. Part of the goal of GapShuffler is for text generation, where using the same word in rapid succession makes the writing look less "educated." It can also be useful for color selection; in pixel art, using the same colors for skin and for long hair will make a human look like they have tentacles on their head, so you'd almost always want different colors for those two.

tommyettinger commented on ShaiRandom/GapShuffler.cs:
```diff
@@ -0,0 +1,283 @@
+        /// This iterator does not support resetting.
+        /// </summary>
+        /// <exception cref="NotSupportedException"/>
+        void IEnumerator.Reset() => throw new NotSupportedException();
```
This doesn't bother me, but is there a technical reason why Reset() can't be implemented? I can see the case being made that because Reset() doesn't take an IEnhancedRandom or its seed, it can't actually reset the GapShuffler back to its starting state. Since most `IEnumerator`s should reset completely when Reset() is called, and this can't, not implementing may the best option.

Chris3606 replied:
Resetting the RNG is well beyond the scope of what the enumerator could reasonable do (or probably what it _should_ do, since it has no idea if the RNG is shared), yeah.  The biggest problem, though, isn't the RNG, it's the array itself.  I can't restore the array to the original ordering, since I have no idea what that original ordering is or how it compares to the original.

Reset isn't really used too much in most of the common use cases for enumerators, so I don't think this is too much of an issue.  In either case, I'm not sure how I could solve the problem of not knowing the original ordering of the list.

tommyettinger commented on ShaiRandom/GapShuffler.cs:
```diff
@@ -0,0 +1,283 @@
+    /// to get an instance of this, rather than creating one yourself.
+    /// </summary>
+    /// <remarks>
+    /// This enumerator is infinite, so you will need to use it in a foreach loop and break out of it, use something
+    /// like LINQ's .Take(n) function, or use it in a while loop with a condition to break out of it in order to prevent
+    /// an infinite loop.  Until you break out of the loop, it will continue to produce a mostly random shuffled stream
+    /// of the items. These shuffles are spaced so that a single element should always have a large amount of "gap" in
```
The way I usually use GapShuffler in Java code is actually not in a loop, but by calling `.next()` when the next single item is needed. Is there an equivalent API here? It might be a default method in one of the enumerating-related interfaces.

Chris3606 replied:
Similar to how you use MoveNext in an Enumerator by hand:

```CS
var shuffler = rng.GapShuffle(list);
shuffler.MoveNext();
Console.WriteLine(shuffler.Current);
```

It's kind of clunky, but that is how C# defines the IEnumerator interface.  You could of course add a `bool Next(out TItem item)` to make it a bit easier, which might not be a bad idea here.

Chris3606 replied:
Actually the helper should really just be `void Next()` since the boolean will always be true (these enumerators are infinite).

tommyettinger commented on ShaiRandom/GapShuffler.cs:
```diff
@@ -0,0 +1,283 @@
+        /// </summary>
+        /// <param name="rng">RNG to use for shuffling.</param>
+        /// <param name="items">Items to shuffle, which are copied into an array internally.</param>
+        public GapShufflerEnumerator(IEnhancedRandom rng, ReadOnlySpan<TItem> items)
```
It's a little funny that this can mutate a `ReadOnlySpan<TItem>`, though I'm not complaining. I guess they aren't really read-only!

Chris3606 replied:
It doesn't mutate it, it copies it into an array; so it never mutates the span, it mutates the copy :D.  The reason I didn't use `Span` here is because, I _can't_ store a Span inside a struct; it's a ref struct, which functionally means it's allowed to exist on the stack and that's about it; can't use them as fields, etc; so there is no nice way to implement an in-place version of this on a Span.   Really this is what Memory<T> is for, thus the other iterator that uses `Memory<T>`

tommyettinger reviewed (approved):
The code looks clean, as usual! I'm a little uncertain about the recommended ways to use this, since they are different from how I usually employ this type of code. As long as there's some easy-enough way to get the next single item without requiring a loop, I'm totally satisfied.

Chris3606 committed:
Implemented Next() functions to allow users to easily manually go through an iterator.

Chris3606 committed:
Renamed functions to use "shuffler", the verb.

Chris3606 committed:
Refactored documentation to better describe the algorithm and note the common use case of advancing the enumerator manually via Next.

Chris3606 committed:
Adding better unit tests.

Chris3606 committed:
Finished improving unit tests and avoided some unneeded code duplication.

Chris3606 commented:
The comments should be addressed, I made the following changes:
1. Added a `.Next()` function to the enumerators make manual advancement very easy.
2. Refactored the documentation to use your comments/explanations, and more directly imply the common use case of simply using `Next()`

I don't know a way to reasonably address the `Reset()` portion; I think it may just have to remain unsupported.  Per the documentation for [IEnumerator.Reset](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator.reset?view=net-7.0#remarks):

> The [Reset](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator.reset?view=net-7.0) method is provided for COM interoperability. It does not necessarily need to be implemented; instead, the implementer can simply throw a [NotSupportedException](https://learn.microsoft.com/en-us/dotnet/api/system.notsupportedexception?view=net-7.0).

So, I don't think users will run into cases involving Reset for this; and I think given that the initial state of the array is different either way, it makes sense to instead force the user to re-create the object (and since it's a struct, this has minimal overhead).

I also finished/improve the unit tests.  They're still a little hacky in spots, but it's at least documented why; feel free to let me know what you think!

tommyettinger commented on ShaiRandom.UnitTests/GapShufflerTests.cs:
```diff
@@ -0,0 +1,156 @@
+                Assert.Equal(originalList.ToHashSet(), half.ToHashSet());
+
+                // No set is the original list.  This is not _technically_ enforced by the algorithm, but it's
+                // highly unlikely to happen by chance (and won't happen with the particular RNG seed we're using).
```
This is definitely a little hacky, but since MizuchiRandom should be stable and the code to shuffle or get bounded ints should be too, relying on the seed being one of the vast majority of seeds that have this behavior should be fine.

tommyettinger commented:
I think this looks good! Thanks for incorporating my quick summary in the docs; I hope it helps people find good usages for this. Having Next() should make it much easier to port Thesaurus from SquidLib/SquidSquad, since that class uses Next()-style advancement heavily. I'll merge this now, big thanks!
